bee-network/src/service/service.rs

```diff
@@ -0,0 +1,486 @@
+            error!("Error processing internal event. Cause: {}", e);
+            continue;
+        }
+    }
+
+    debug!("Event processor stopped.");
+}
+
+// TODO: implement exponential back-off to not spam the peer with reconnect attempts.
+async fn peer_checker(shutdown: Shutdown, senders: Senders, peerlist: PeerList) {
+    debug!("Peer checker running.");
+
+    let Senders { internal_commands, .. } = senders;
+
+    // NOTE:
+    // We add a random amount of milliseconds to when the reconnector starts, so that even if 2 nodes
+    // go online at the same time the probablilty of them simultaneously dialing each other is reduced
+    // significantly.
+
+    let delay = Duration::from_millis(rand::thread_rng().gen_range(0u64..1000));
+    let start = Instant::now() + delay;
+    let period = Duration::from_secs(reconnect_interval_secs()); // `unwrap` is safe!
+
+    let mut interval = ShutdownStream::new(shutdown, IntervalStream::new(time::interval_at(start, period)));
+
+    // Check, if there are any disconnected known peers, and schedule a reconnect attempt for each
+    // of those.
+    while interval.next().await.is_some() {
+        let peerlist = peerlist.0.read().await;
+
+        for (peer_id, alias) in peerlist.iter_if(|info, state| info.relation.is_known() && state.is_disconnected()) {
+            info!("Trying to reconnect to: {} ({}).", alias, alias!(peer_id));
+
+            // Ignore if the command fails. We can always try another time.
+            let _ = internal_commands.send(Command::DialPeer { peer_id });
+        }
+    }
+
+    debug!("Peer checker stopped.");
+}
+
+async fn process_command(command: Command, senders: &Senders, peerlist: &PeerList) -> Result<(), PeerError> {
+    trace!("Received {:?}.", command);
+
+    match command {
+        Command::AddPeer {
+            peer_id,
+            multiaddr,
+            alias,
+            relation,
+        } => {
+            let alias = alias.unwrap_or_else(|| alias!(peer_id).to_string());
+
+            // Note: the control flow seems to violate DRY principle, but we only need to clone `id` in one branch.
+            if relation.is_known() {
+                add_peer(peer_id, multiaddr, alias, relation, senders, peerlist).await?;
+
+                // We automatically connect to such peers. Since we can connect concurrently, we spawn a task here.
+                let _ = senders.internal_commands.send(Command::DialPeer { peer_id });
+            } else {
+                add_peer(peer_id, multiaddr, alias, relation, senders, peerlist).await?;
+            }
+        }
+
+        Command::BanAddress { address } => {
+            let mut peerlist = peerlist.0.write().await;
+            peerlist.ban_address(address.clone())?;
+
+            let _ = senders.events.send(Event::AddressBanned { address });
+        }
+
+        Command::BanPeer { peer_id } => {
+            let mut peerlist = peerlist.0.write().await;
```

Comment:
Creating the guard on a separate line here holds mutable access to the peer list even over the `.send` below. I assume this is unintentional. Do it all on one line (i.e: `peerlist.0.write().await.ban_peer(peer_id)?` to shrink the critical section. The same applies to other cases in this file.

Reply:
good point. will change

Reply:
I've not noted it, but this pattern appears in several files in the PR.

Reply:
Changed it in all places in this file, and where I couldn't one-line this, made sure to drop the lock before the `send` happens.